src/python/pants/engine/legacy/source_mapper.py

```diff
@@ -48,20 +49,27 @@ def _unique_dirs_for_sources(self, sources):
   def target_addresses_for_source(self, source):
     return list(self.iter_target_addresses_for_sources([source]))
 
-  def _iter_owned_files_from_hydrated_target(self, legacy_target):
+  def _match_source(self, source, fileset):
+    if os.path.exists(source):
```

Comment:
Rather than always checking whether it exists, maybe first check whether it is in the fileset... if it is, then it's definitely owned, and that's cheaper than a syscall. If it's not in the fileset, can use `matches_filespec`.